### PR TITLE
refactor(connlib): split device handler for clients & gateway

### PR DESCRIPTION
**Pull request by thomaseizinger**
This PR splits the device handler into dedicated functions on `Tunnel` that are parameterized with `ClientState` and `GatewayState`. I didn't understand all the inner workings so I tried to move very carefully and make small commits. Review patch-by-patch is recommended. All commits pass `cargo clippy`.

This is in preparation for removing `ControlSignal` which is only used on the client because the gateway never sends connection intents. To achieve that, I needed to have a dedicated device handler for the gateway that does not end up calling `connection_intent` so that one could be moved to the `ClientState` as a result.

This PR does not yet remove `ControlSignal` because that is in itself a big task and I wanted to verify that I am doing all the right things here first.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
@conectado This currently fails with a stack overflow in the tokio runtime. I think that is because I moved the buffers around. I am surprised that using stack-allocated buffers of this size makes things faster. I would have deemed two heap allocations per network device to be not noticable?

Curious to get your thoughts in this. Embedding the read buffer results in a much nicer API. Do you think we can get away with allocating that one on the heap?

File: rust/connlib/tunnel/src/lib.rs [hunk 1]
```diff
@@ -554,3 +505,244 @@ where
...
+        tracing::debug!("background_loop_started");
+
+        Ok(())
+    }
+
+    fn start_device(self: &Arc<Self>, mut device: Device) {
+        let tunnel = Arc::clone(self);
+
+        *self.iface_handler_abort.lock() = Some(
+            tokio::spawn(async move {
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: For both `start_device` I'd like if we can move closure to its own function? I find it easier to read.

**Review by conectado — APPROVED** [review]
Awesome work!

I looked at it commit by commit and I think every change is correct.

Just want to make sure integration tests run before merging.

Also, what do you think of moving the gateway/client specific `impl` blocks to their own module? I think that will make it easier to navigate. 

And lastly, just in case, `peer_handler` also have gateway/client specific functions so I think we should do similar changes there.

**Comment by conectado** [comment]
> @conectado This currently fails with a stack overflow in the tokio runtime. I think that is because I moved the buffers around. I am surprised that using stack-allocated buffers of this size makes things faster. I would have deemed two heap allocations per network device to be not noticable?
> 
> Curious to get your thoughts in this. Embedding the read buffer results in a much nicer API. Do you think we can get away with allocating that one on the heap?

Just read this after my review.

I think that's fine as long as we make sure to not re-alloc on each packet read, for that purpose I'd use a `Box<[u8; MAX_UDP_SIZE]>` rather than a `Vec`.

**Comment by thomaseizinger** [comment]
> Also, what do you think of moving the gateway/client specific `impl` blocks to their own module? I think that will make it easier to navigate.

Will do together with fixing the integration tests and then merge!

> And lastly, just in case, `peer_handler` also have gateway/client specific functions so I think we should do similar changes there.

Okay, sounds good!

Just to double check: The way I removed functionality from the client and gateway handler regarding DNS is correct?

**Comment by conectado** [comment]
> > Also, what do you think of moving the gateway/client specific `impl` blocks to their own module? I think that will make it easier to navigate.
> 
> Will do together with fixing the integration tests and then merge!
> 
> > And lastly, just in case, `peer_handler` also have gateway/client specific functions so I think we should do similar changes there.
> 
> Okay, sounds good!
> 
> Just to double check: The way I removed functionality from the client and gateway handler regarding DNS is correct?

Yeah, I checked that during the review :)

**Pull request merged**